DataStructures/MinMaxHeap.cs

```diff
@@ -0,0 +1,399 @@
+            const uint minLevelsBits = 0x55555555;
+            const uint maxLevelsBits = 0xAAAAAAAA;
+            return ((index + 1) & minLevelsBits) > ((index + 1) & maxLevelsBits);
+        }
+
+        private int Parent(int index) => (index - 1) / 2;
+
+        private int Grandparent(int index) => ((index - 1) / 2 - 1) / 2;
+
+        /// <summary>
+        /// Assuming that children sub-trees are valid heaps, pushes node to lower levels
+        /// to make heap valid.
+        /// </summary>
+        /// <param name="index">Node index.</param>
+        private void PushDown(int index)
+        {
+            if (IsMinLevelIndex(index))
+            {
+                PushDownMin(index);
+            }
+            else
+            {
+                PushDownMax(index);
+            }
+        }
+
+        private void PushDownMax(int index)
+        {
+            if (!HasChild(index))
+            {
+                return;
+            }
+
+            int maxIndex = ExtremumChildOrGrandchild(index, minimum: false);
+
+            // If smaller element are put at min level (as result of swaping), it doesn't affect sub-tree validity.
+            // If smaller element are put at max level, PushDownMax() should be called for that node.
+            if (IsGrandchild(index, maxIndex))
+            {
+                if (Comparer.Compare(Heap[maxIndex], Heap[index]) > 0)
+                {
+                    SwapNodes(maxIndex, index);
+                    if (Comparer.Compare(Heap[maxIndex], Heap[Parent(maxIndex)]) < 0)
+                    {
+                        SwapNodes(maxIndex, Parent(maxIndex));
+                    }
+
+                    PushDownMax(maxIndex);
+                }
+            }
+            else if (Comparer.Compare(Heap[maxIndex], Heap[index]) > 0)
+            {
+                SwapNodes(maxIndex, index);
+            }
+        }
+
+        private void PushDownMin(int index)
+        {
+            if (!HasChild(index))
+            {
+                return;
+            }
+
+            int minIndex = ExtremumChildOrGrandchild(index, minimum: true);
+
+            // If bigger element are put at max level (as result of swaping), it doesn't affect sub-tree validity.
+            // If bigger element are put at min level, PushDownMin() should be called for that node.
+            if (IsGrandchild(index, minIndex))
+            {
+                if (Comparer.Compare(Heap[minIndex], Heap[index]) < 0)
+                {
+                    SwapNodes(minIndex, index);
+                    if (Comparer.Compare(Heap[minIndex], Heap[Parent(minIndex)]) > 0)
+                    {
+                        SwapNodes(minIndex, Parent(minIndex));
+                    }
+
+                    PushDownMin(minIndex);
+                }
+            }
+            else if (Comparer.Compare(Heap[minIndex], Heap[index]) < 0)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add the missing 'else' clause.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5421646)